feat(GraphList): Add numEdges functionality to graph storage

Ruairia opened the pull request:
## Which issue does this PR close?

<!--
We generally require a GitHub issue to be filed for all bug fixes and enhancements and this helps us generate change logs for our releases. You can link an issue to this PR using the GitHub syntax. For example `Closes #123` indicates that this PR will close issue #123.
-->

- Closes #77 .

## Rationale for this change
This PR addresses the feature request to add support for retrieving the total number of edges in graph storage. Calculating edge count manually instead of using previously calculated metadata provides better accuracy.
<!--
 Why are you proposing this change? If this is already explained clearly in the issue then this section is not needed.
 Explaining clearly why changes are proposed helps reviewers understand your changes and offer better suggestions for fixes.  
-->

## What changes are included in this PR?
- Add impl_numEdges() pure virtual method to PeakStorageInterface
- Implement impl_numEdges() in AdjacencyList (counts by iterating adjacency lists)
- Implement impl_numEdges() in HybridCSR_COO (counts COO format edges)
- Add numEdges() method to PeakStore (delegates to active storage)
- Add numEdges() method to GraphList (provides public API)
- Add comprehensive test suite with 7 test cases covering various scenarios
- All tests pass successfully
<!--
There is no need to duplicate the description in the issue here but it is sometimes worth providing a summary of the individual changes in this PR.
-->

## Are these changes tested?
Yes.
All new functionality is covered by NumEdgesTest.cpp (7 passing test cases).
Verified correctness incrementally (edge count updates as edges are added).
Confirmed all existing tests pass without regressions.
<!--
We typically require tests for all PRs in order to:
1. Prevent the code from being accidentally broken by subsequent changes
2. Serve as another way to document the expected behavior of the code

If tests are not included in your PR, please explain why (for example, are they covered by existing tests)?
-->

## Are there any user-facing changes?
Yes.
GraphList now exposes a new public method numEdges() to efficiently retrieve the total edge count.
This is an additive, non-breaking change that does not affect existing functionality.
<!--
If there are user-facing changes then we may require documentation to be updated before approving the PR.
-->

<!--
If there are any breaking changes to public APIs, please add the `api change` label.
-->

Commit:
Added functions to enable/disable log display

Commit:
Examples updated to test toggling functionality

Commit:
log toggling functionality tested successfully

Commit:
renamed method name togglePLogging to setConsoleLogging

SharonIV0x86 committed:
Merge branch 'main' into logging

Ruairia committed:
Add numEdges functionality to graph storage

- Add impl_numEdges() pure virtual method to PeakStorageInterface
- Implement impl_numEdges() in AdjacencyList (counts by iterating adjacency lists)
- Implement impl_numEdges() in HybridCSR_COO (counts COO format edges)
- Add numEdges() method to PeakStore (delegates to active storage)
- Add numEdges() method to GraphList (provides public API)
- Add comprehensive test suite with 7 test cases covering various scenarios
- All tests pass successfully

Ruairia committed:
# Conflicts:
#	src/GraphList.hpp
#	src/PeakStore.hpp

SharonIV0x86 commented on tests/NumEdgesTest.cpp:
```diff
@@ -0,0 +1,138 @@
+//
+// Created by ruairi on 05/09/2025.
+//
```
It would be better if you could move your tests to this file. And try to cover all possible cases for ``numEdges`` under 2-3 test cases not more than that.
[AdjacencyList.hpp](https://github.com/SharonIV0x86/CinderPeak/blob/main/tests/AdjacencyShard.cpp)

SharonIV0x86 commented on src/StorageEngine/AdjacencyList.hpp:
```diff
@@ -177,6 +177,15 @@ class AdjacencyList
+  size_t impl_numEdges() const override
+  {
+    size_t num_edges = 0;
+    for (const auto &[first, second] : _adj_list) {
```
Instead of traversing the ``_adj_list`` which can be a expensive operation, try to see if you can use properties of ``GraphInternalMetadata``.

Ruairia replied:
May I ask for some clarification on the best architectural approach.
Current situation:

PeakStore::addEdge() already tracks edge count via ctx->metadata->num_edges++
As you pointed out, the expensive O(V+E) traversal in AdjacencyList::impl_numEdges() is redundant since we already have the count

Two implementation options:
Option A: Metadata-based (O(1))

```c++
// In PeakStore.hpp
size_t numEdges() const {
    return ctx->metadata->num_edges;  // Direct metadata access
}

``` 

Option B: Pass metadata to AdjacencyList (still O(1))

```c++
// Modify AdjacencyList constructor to accept metadata
// Then use metadata within AdjacencyList::impl_numEdges()
size_t impl_numEdges() const override {
    return _metadata->num_edges;
}

``` 

Which approach aligns better with the codebase architecture?
Option A is simpler since the metadata is already being maintained at the PeakStore level, but Option B follows the original issue's request to implement impl_numEdges() in AdjacencyList.hpp.

SharonIV0x86 replied:
This did cross my mind, as per the design the ``PeakStore`` does the management of metadata. edge count belongs to the store, and ``AdjacencyList`` is just a detail of representation used for storage and querying.
So there is no need to keep the ``impl_numEdges`` in ``AdjacencyList.hpp`` and ``HybridCSR_COO.hpp`` as it can be handled in the PeakStore

Ruairia replied:
Updated numEdges() to use metadata directly instead of delegating to storage engines.
Achieves O(1) performance and keeps storage engines focused only on representation.
Also moved tests to Statistics.cpp and condensed from 7 into 3.

SharonIV0x86 reviewed (changes requested):
Need changes.

SharonIV0x86 committed:
Update CONTRIBUTING.md

SharonIV0x86 committed:
Merge pull request #67 from me-hem/logging

feat(logging): Added setConsoleLogging method to enable or disable logs in terminal.

Ruairia committed:
Moved test cases to Statistics.cpp

Ruairia committed:
Moved test cases to Statistics.cpp

Ruairia committed:
Merge remote-tracking branch 'origin/Implement_numEdges()' into Implement_numEdges()

# Conflicts:
#	src/GraphList.hpp

Ruairia committed:
Removed duplicate function definition

Ruairia committed:
feat: optimize numEdges() using metadata tracking

- Remove impl_numEdges() from storage engines (AdjacencyList, HybridCSR_COO)
- Remove impl_numEdges() from PeakStorageInterface
- Update PeakStore::numEdges() to use ctx->metadata->num_edges directly
- Achieves O(1) performance by leveraging existing edge count tracking
- Maintains clean separation: PeakStore handles graph semantics, storage engines handle representation

Resolves feedback to use GraphInternalMetadata properties instead of expensive traversal operations.

SharonIV0x86 commented:
@Ruairia This is weird, i can see the changes which were already merged by existing PR, if possible close this PR and open a new one with only mentioned changes as most of the commits you made are for fixing the conflicts, see if you can open a new PR or make changes in the existing one.

Ruairia commented:
New Pull request opened, are you able to see it?
